emasamplerate.go

```diff
@@ -0,0 +1,300 @@
+	// the EMA. With larger values, newer data will influence the average more, and older
+	// values will be factored out more quickly.  In mathematical literature concerning EMA,
+	// this is referred to as the `alpha` constant.
+	// Default is 0.5
+	Weight float64
+
+	// GoalSampleRate is the average sample rate we're aiming for, across all
+	// events. Default 10
+	GoalSampleRate int
+
+	// MaxKeys, if greater than 0, limits the number of distinct keys tracked in EMA.
+	// Once MaxKeys is reached, new keys will not be included in the sample rate map, but
+	// existing keys will continue to be be counted.
+	MaxKeys int
+
+	// AgeOutValue indicates the threshold for removing keys from the EMA. The EMA of any key will approach 0
+	// if it is not repeatedly observed, but will never truly reach it, so we have to decide what constitutes "zero".
+	// Keys with averages below this threshold will be removed from the EMA. Default is the same as Weight, as this prevents
+	// a key with the smallest integer value (1) from being aged out immediately.
+	AgeOutValue float64
+
+	// BurstMultiple, if set, is multiplied by the sum of the running average of counts to define
+	// the burst detection threshold. If total counts observed for a given interval exceed the threshold
+	// EMA is updated immediately, rather than waiting on the AdjustmentInterval.
+	// Defaults to 2; negative value disables. With a default of 2, if your traffic suddenly doubles,
+	// burst detection will kick in.
+	BurstMultiple float64
+
+	// BurstDetectionDelay indicates the number of intervals to run after Start is called before burst detection kicks in.
+	// Defaults to 3
+	BurstDetectionDelay uint
+
+	savedSampleRates map[string]int
+	currentCounts    map[string]float64
+	movingAverage    map[string]float64
+	burstThreshold   float64
+	currentBurstSum  float64
+	intervalCount    uint
+	burstSignal      chan struct{}
+
+	// haveData indicates that we have gotten a sample of traffic. Before we've
+	// gotten any samples of traffic, we should we should use the default goal
+	// sample rate for all events instead of sampling everything at 1
+	haveData bool
+
+	lock sync.Mutex
+
+	// used only in tests
+	testSignalMapsDone chan struct{}
+}
+
+func (e *EMASampleRate) Start() error {
+	// apply defaults
+	if e.AdjustmentInterval == 0 {
+		e.AdjustmentInterval = 15
+	}
+	if e.GoalSampleRate == 0 {
+		e.GoalSampleRate = 10
+	}
+	if e.Weight == 0 {
+		e.Weight = 0.5
+	}
+	if e.AgeOutValue == 0 {
+		e.AgeOutValue = e.Weight
+	}
+	if e.BurstMultiple == 0 {
+		e.BurstMultiple = 2
+	}
+	if e.BurstDetectionDelay == 0 {
+		e.BurstDetectionDelay = 3
+	}
+
+	e.savedSampleRates = make(map[string]int)
+	e.currentCounts = make(map[string]float64)
+	e.movingAverage = make(map[string]float64)
+	e.burstSignal = make(chan struct{})
+
+	go func() {
+		ticker := time.NewTicker(time.Second * time.Duration(e.AdjustmentInterval))
+		for {
+			select {
```

Comment:
I think the burst signal and the ticker will be able to race and both attempt to update maps simultaneously, introducing some thread unsafety around modifying the moving average map. If the ticker fires and then the burst fires immediately after, they'll both contend for the `e.lock` to make their own local copy of the current counts. But say they both finish doing that, they'll both try and run `e.updateEMA()` simultaneously and the concurrent writes will panic.

Reply:
Good catch.

Reply:
This should be fixed in https://github.com/honeycombio/dynsampler-go/pull/10/commits/a0d1840156ae33409d4533b33e256bfd06416cdb - added a new `updating` variable to prevent multiple calls to `updateMaps` from conflicting (and doing double work).